CSharp/Library/QnAMaker/QnAMakerDialog.cs

```diff
@@ -1,110 +1,225 @@
-
-                var maxValue = tasks.Max(x => x.Result.Score);
-                qnaMakerResult = await tasks.First(x => x.Result.Score == maxValue);
-
-                await this.RespondFromQnAMakerResultAsync(context, message, qnaMakerResult);
-            }
-
-            await this.DefaultWaitNextMessageAsync(context, message, qnaMakerResult);
-        }
-
-        protected virtual async Task RespondFromQnAMakerResultAsync(IDialogContext context, IMessageActivity message, QnAMakerResult result)
-        {
-            result.Score /= 100;
-            var answer = result.Score >= result.ServiceCfg.ScoreThreshold ? result.Answer : result.ServiceCfg.DefaultMessage;
-
-            await context.PostAsync(HttpUtility.HtmlDecode(answer));
-        }
-
-        protected virtual async Task DefaultWaitNextMessageAsync(IDialogContext context, IMessageActivity message, QnAMakerResult result)
-        {
-            context.Wait(MessageReceivedAsync);
-        }
-    }
-}
+﻿// 
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license.
+// 
+// Microsoft Bot Framework: http://botframework.com
+// 
+// Cognitive Services based Dialogs for Bot Builder:
+// https://github.com/Microsoft/BotBuilder-CognitiveServices 
+// 
+// Copyright (c) Microsoft Corporation
+// All rights reserved.
+// 
+// MIT License:
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED ""AS IS"", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.Dialogs;
+using Microsoft.Bot.Connector;
+using Microsoft.Bot.Builder.Internals.Fibers;
+using System.Reflection;
+using System.Linq;
+using System.Web;
+
+namespace Microsoft.Bot.Builder.CognitiveServices.QnAMaker
+{
+    /// <summary>
+    /// A dialog specialized to handle QnA response from QnA Maker.
+    /// </summary>
+    [Serializable]
+    public class QnAMakerDialog : IDialog<IMessageActivity>

```

Comment:
>QnAMakerDialog [](start = 17, length = 14)

can you add tests with some mocked QnA services for both QnAMakerDialog and QnAMakerScorable?